src/ensembl/src/content/app/browser/hooks/useGenomeBrowser.ts

```diff
@@ -0,0 +1,186 @@
+import { BROWSER_CONTAINER_ID } from 'src/content/app/browser/browser-constants';
+
+import browserStorageService from 'src/content/app/browser/browser-storage-service';
+
+import {
+  getBrowserActiveGenomeId,
+  getBrowserActiveEnsObjectId
+} from 'src/content/app/browser/browserSelectors';
+import { updatePreviouslyViewedObjectsAndSave } from 'src/content/app/browser/track-panel/trackPanelActions';
+
+import { GenomeBrowserContext } from 'src/content/app/browser/Browser';
+import { TrackStates } from 'src/content/app/browser/track-panel/trackPanelConfig';
+import { Status } from 'src/shared/types/status';
+import { ChrLocation } from 'src/content/app/browser/browserState';
+import { parseEnsObjectId } from 'src/shared/state/ens-object/ensObjectHelpers';
+import { getChrLocationFromStr } from 'src/content/app/browser/browserHelper';
+
+const useGenomeBrowser = () => {
+  const dispatch = useDispatch();
+
+  const activeGenomeId = useSelector(getBrowserActiveGenomeId);
+  const activeEnsObjectId = useSelector(getBrowserActiveEnsObjectId);
+
+  const { genomeBrowser, setGenomeBrowser, setZmenus, zmenus } =
+    useContext(GenomeBrowserContext);
+
+  const restoreBrowserTrackStates = () => {
+    if (!activeGenomeId || !activeEnsObjectId || !genomeBrowser) {
+      return;
+    }
+
+    const trackStatesFromStorage = browserStorageService.getTrackStates();
+    const mergedTrackStates = {
+      ...get(
+        trackStatesFromStorage,
+        `${activeGenomeId}.objectTracks.${activeEnsObjectId}`
+      ),
+      ...get(trackStatesFromStorage, `${activeGenomeId}.commonTracks`)
+    } as TrackStates;
+
+    const tracksToTurnOff: string[] = [];
+    const tracksToTurnOn: string[] = [];
+
+    Object.values(mergedTrackStates).forEach((trackStates) => {
+      Object.keys(trackStates).forEach((trackId) => {
+        const track_id =
+          trackId === 'track:gene-feat' || trackId === 'track:gene-feat-1'
+            ? 'focus'
+            : trackId.replace('track:', '');
+
+        trackStates[trackId] === Status.SELECTED
+          ? tracksToTurnOn.push(track_id)
+          : tracksToTurnOff.push(track_id);
+      });
+    });
+
+    if (tracksToTurnOn.length) {
+      const turnOnAction: OutgoingAction = {
+        type: OutgoingActionType.TURN_ON_TRACKS,
+        payload: {
+          track_ids: tracksToTurnOn
+        }
+      };
+      genomeBrowser.send(turnOnAction);
+    }
+
+    if (tracksToTurnOff.length) {
+      const turnOffAction: OutgoingAction = {
+        type: OutgoingActionType.TURN_OFF_TRACKS,
+        payload: {
+          track_ids: tracksToTurnOff
+        }
+      };
+
+      genomeBrowser.send(turnOffAction);
+    }
+  };
+
+  const activateGenomeBrowser = async () => {
+    const genomeBrowserService = new EnsemblGenomeBrowser();
+    await genomeBrowserService.init({
+      backend_url: config.genomeBrowserBackendBaseUrl,
+      target_element_id: BROWSER_CONTAINER_ID
+    });
+    if (setGenomeBrowser) {
```

Comment:
Ok, I can see now that the reason `setGenomeBrowser` is considered to be nullable is because of the default value of the context.

There are three options for dealing with this.

1. One is to set every field on the context type as nullable, as done in this PR.
2. Set non-nullable fields to some default values, e.g. the noop function (discussed during code review)
3. The most interesting option: describe the type of the context as you expect it to be, and create the context with the type `GenomeBrowserContext | undefined`. Then, in the `useGenomeBrowser` hook, retrieve the context, make sure it is not `undefined`, and use it normally.

This third option is described in a blog post by Kent C. Dodds [here](https://kentcdodds.com/blog/how-to-use-react-context-effectively#typescript).

For our case, that would be:

```
const GenomeBrowserContext = React.createContext<GenomeBrowserContextType | undefined>(undefined);

// then in the consuming hook

const genomeBrowserContext = useContext(GenomeBrowserContext);
if (!genomeBrowserContext) {
  throw new Error('useGenomeBrowser must be used with GenomeBrowserContext Provider');
}

// at this point, the context type is GenomeBrowserContextType
do_stuff()
```